### PR TITLE
DP-11: Remove monitor messages in unmonitor

### PR DESCRIPTION
**Pull request by VicNumber21**

### CODE REVIEW

**Comment by facundominguez** [comment]
Thanks, Victor. Which problem are you trying to address with this patch? I'm making a guess, but I don't think the patch solves it.

**Comment by VicNumber21** [comment]
@facundominguez you're welcome.

I've tried to remove messages from message queue to prevent catching them after unmonitor called.
I found the task description here https://cloud-haskell.atlassian.net/projects/DP/issues/DP-11?filter=allopenissues and decided to try to address it.

May you please explain why it is not solution for the problem? I'll rework if required.

**Comment by facundominguez** [comment]
We have moved tickets from Jira to github. Here is the github one:
https://github.com/haskell-distributed/distributed-process/issues/55

I'm not sure I make sense of the problem. Will discuss it there.

**Comment by VicNumber21** [comment]
I see. Probably it is required to update http://haskell-distributed.github.io/ (Issues link at the bottom) and http://haskell-distributed.github.io/wiki/contributing.html (point 1, link to Jira there) to remove reference to Jira.
It may help to newcomers to get things right.

I'll track the discussion in the issue and wait for your decision.

**Comment by mboes** [comment]
@facundominguez any closure on #55 and what to do with this PR?

File: src/Control/Distributed/Process/Internal/Primitives.hs [hunk 1]
```diff
@@ -917,9 +919,17 @@ unlinkPort sport = do
 unmonitor :: MonitorRef -> Process ()
 unmonitor ref = do
   unmonitorAsync ref
-  receiveWait [ matchIf (\(DidUnmonitor ref') -> ref' == ref)
-                        (\_ -> return ())
-              ]
+  receiveWait [ matchIf isDidUnmonitor ignore ]
+  receiveAllNow [ matchIf isProcessMonitorNotification ignore,
```
Comment (facundominguez) [inline src/Control/Distributed/Process/Internal/Primitives.hs]: I guess there is no need to collect a list of units `[()]` as receiveAllNow was defined to do below. How about writing this in a single loop here:
```
fix $ \loop ->
  m <- receiveTimeout 0 [...]
  forM_ m loop
```

File: src/Control/Distributed/Process/Internal/Primitives.hs [hunk 2]
```diff
@@ -917,9 +919,17 @@ unlinkPort sport = do
...
+                  matchIf isNodeMonitorNotification ignore,
+                  matchIf isPortMonitorNotification ignore
+                ]
+  return ()
+    where ignore _ = return ()
+          isDidUnmonitor (DidUnmonitor ref') = ref' == ref
+          isProcessMonitorNotification (ProcessMonitorNotification ref' _ _) = ref' == ref
+          isNodeMonitorNotification (NodeMonitorNotification ref' _ _) = ref' == ref
+          isPortMonitorNotification (PortMonitorNotification ref' _ _) = ref' == ref
```
Comment (facundominguez) [inline src/Control/Distributed/Process/Internal/Primitives.hs]: Please, inline all of these local definitions.

File: src/Control/Distributed/Process/Internal/Primitives.hs [hunk 3]
```diff
@@ -1003,6 +1013,25 @@ catchesHandler handlers e = foldr tryHandler (throw e) handlers
 expectTimeout :: forall a. Serializable a => Int -> Process (Maybe a)
 expectTimeout n = receiveTimeout n [match return]
 
+-- | Synonim of 'receiveTimeout' 0.
+--
+-- Immediatelly take a matched message from the message queue.
+receiveNow :: [Match b] -> Process (Maybe b)
+receiveNow = receiveTimeout 0
```
Comment (facundominguez) [inline src/Control/Distributed/Process/Internal/Primitives.hs]: Please, inline this at the (only) call site.

File: src/Control/Distributed/Process/Internal/Primitives.hs [hunk 4]
```diff
@@ -917,9 +919,17 @@ unlinkPort sport = do
 unmonitor :: MonitorRef -> Process ()
```
Comment (facundominguez) [inline src/Control/Distributed/Process/Internal/Primitives.hs]: Please, update the documentation to explain how the mailbox is affected.

**Comment by facundominguez** [comment]
@VicNumber21, I just did a review pass making some observations. Also, could you add a mention of #55 in the commit message?

File: src/Control/Distributed/Process/Internal/Primitives.hs (hunk 1 above)
Comment (facundominguez) [inline src/Control/Distributed/Process/Internal/Primitives.hs]: Actually, it looks to me like there can be at most one monitor notification in the mailbox for a given `MonitorRef`. Therefore no loop would be necessary. It should suffice:
```Haskell
_ <- receiveTimeout 0 [ ... ]
```

**Comment by facundominguez** [comment]
Would be merging as:
https://github.com/haskell-distributed/distributed-process/pull/268

**Pull request closed without merging**